polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -703,13 +710,218 @@ void AffineParallelGuardBuilder::getConstraints(
     }
 }
 
+//===----------------------------------------------------------------------===//
+// SYCLAccessorVersionBuilder
+//===----------------------------------------------------------------------===//
+
+// TODO: Support other kind of loops.
+class SYCLAccessorVersionBuilder : public SCFLoopVersionBuilder {
+public:
+  SYCLAccessorVersionBuilder(LoopLikeOpInterface loop)
+      : SCFLoopVersionBuilder(loop) {}
+
+  bool isVersionCandidate(const Operation &op1, const Operation &op2) {
+    if (!EnableSYCLAccessorVersioning)
+      return false;
+
+    // Currently only allow one accessor pair.
+    if (!empty())
+      return false;
```

Comment:
Maybe assert?

Reply:
This is a temporary limitation, which only allow one candidate. Either we return false here in `isVersionCandidate`, or we need the user of `isVersionCandidate` to check before calling `isVersionCandidate`, which means exposing `empty()` too. I personally prefer the first way, do you prefer the second way?

Reply:
No, this is good to me :+1: